MarkBind/markbind
# Add support for Octicons

**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [X] New feature

Implements and closes #998. 

**What is the rationale for this request?**
[GitHub's Octicons](https://octicons.github.com/) are a set of icons that represent Git and GitHub-specific features, useful when writing technical documents (about software engineering for example). 

**What changes did you make? (Give an overview)**
- Add support for Octicons (with the `:octicon-(name):` syntax) using the official [Octicons Node library](https://github.com/primer/octicons/tree/master/lib/octicons_node)
- Add unit tests for Octicons
- Update the functional test sites with the Octicon stylesheet (via `npm run updatetest`)
- Update the user guide with examples for Octicon usage

Example: 
<img width="655" alt="Screen Shot 2020-01-27 at 10 26 22 PM" src="https://user-images.githubusercontent.com/6095637/73182470-14c90c80-4154-11ea-89b1-a847c853afac.png">

**Provide some example code that this change will affect:**
```js
// Extends existing icon parsing logic
const octicons = require('@primer/octicons');

module.exports = require('markdown-it-regexp')(
    /:(fa[brs]|glyphicon|octicon)-([a-z-]+):/,
    (match, utils) => {
        let iconFontType = match[1];
        let iconFontName = match[2];

        if (iconFontType === 'glyphicon') {
            return `<span aria-hidden="true" class="glyphicon glyphicon-${iconFontName}"></span>`;
        } else if (iconFontType === 'octicon') {
            return octicons[iconFontName].toSVG();
        } else { // If icon is a Font Awesome icon
            return `<span aria-hidden="true" class="${iconFontType} fa-${iconFontName}"></span>`;
        }
    }
);
```

**Is there anything you'd like reviewers to focus on?**
`Site.js#Site.prototype.copyOcticonsAsset`

The Octicons Node library distributes an unminified stylesheet (identical to [this](https://github.com/primer/octicons/blob/master/lib/octicons_node/index.scss)) that uses 2 spaces for indentation. If the stylesheet is copied directly to the output site (similar to how it's done for Font Awesome and Glyphicons), it causes stylelint errors for the test sites after executing `npm run updatetest` because stylelint expects 4 spaces for indentation. 

Due to this, I had to manually "minify" the CSS file by stripping out all whitespace during the copy. 

**Testing instructions:**
- Use the `:octicon-(name):` syntax to display any Octicon
- Verify that the unit and functional tests are correct, and that `npm run test` is passing

**Proposed commit message: (wrap lines at 72 characters)**
Add support for Octicons
